Rewrite documentation on installing plugins (#8921)
jekyll/jekyll: docs/_docs/plugins/installation.md

permalink: /docs/plugins/installation/
---

Jekyll has built-in support for using plugins to extend the core functionality.

Primarily, any file with extension `.rb` placed within a `_plugins` directory at the root of the site's `source`, will be automatically loaded
during a build session.

This behavior can be configured as follows:

  * The `_plugins` directory may be changed either directly via the command-line or via the configuration file(s).

>   * Plugins in the `_plugins` directory (or its equivalent(s)) will not be loaded when Jekyll is running in `safe` mode.

Reviewer: No indenting need for bullet points

and also dash is the prevalent style for Markdown out level list (`-` then `*` and `+`). Unless the style elsewhere in the docs is to use `*`.
```suggestion
- Plugins in the `_plugins` directory (or its equivalent(s)) will not be loaded when Jekyll is running in `safe` mode.
```

Author: We don't have any fixed style guidelines for Markdown documents.
So I don't have any strong feelings about accepting / rejecting your request..

  * This route cannot be used to extend the Jekyll CLI.

To work with plugins packaged as gems, one has to list the desired gems in the configuration file under a top-level key named `plugins`.
Additionally, if you're building in `safe` mode, the gem needs to be listed under a top-level key named `whitelist`. For example:

```yaml
plugins:
  - jekyll-gist
  - jekyll-coffeescript
  - jekyll-seo-tag
  - some-other-jekyll-plugin

# Enable safe mode
safe: true

# Whitelist plugins under safe mode.
# Note that `some-other-jekyll-plugin` is not listed here. Therefore,
# it will not be loaded under safe mode.
whitelist:
  - jekyll-gist
  - jekyll-coffeescript
  - jekyll-seo-tag
```

In the absence of a Gemfile, one must manually ensure that listed plugins have been installed prior to invoking Jekyll. For example, the
latest versions of gems in the above list may be installed to a system-wide location by running:

```sh
gem install jekyll-gist jekyll-coffeescript jekyll-remote-theme some-other-jekyll-plugin
```

> ## Using a Gemfile

Reviewer: ```suggestion

## Using a Gemfile
```

permalink: /docs/plugins/installation/
---

Jekyll has built-in support for using plugins to extend the core functionality.

Primarily, any file with extension `.rb` placed within a `_plugins` directory at the root of the site's `source`, will be automatically loaded
during a build session.

> This behavior can be configured as per following:

Reviewer 2: "This behavior can be configured as follows:"

  * The `_plugins` directory may be changed either directly via the command-line or via the configuration file(s).

>   * Plugins in the `_plugins` directory (or its equivalent(s)) will not be loaded during the `safe` mode.

Reviewer 3: ```suggestion
  * Plugins in the `_plugins` directory (or its equivalent(s)) will not be loaded when Jekyll is running in `safe` mode.
```

  * This route cannot be used to extend the Jekyll CLI.

> To work with plugins packaged as gems, one has to list the desired gems in the configuration file under a top-level key named `plugins`

Reviewer 3: What do you think about removing the documentation for the `gems` key? I don't think it makes sense to document behavior for anything older than Jekyll 3.9.x.

(or `gems` if your Jekyll version is older than `3.5.0`). Additionally, if you're building under `safe` mode, the gem needs to be listed under

> a top-level key named `whitelist`. For example,

Reviewer 2: "...a top-level key named `whitelist`. For example:"

```yaml
plugins:
  - jekyll-gist
  - jekyll-coffeescript
  - jekyll-seo-tag
  - some-other-jekyll-plugin

# Enable safe mode
safe: true

# Whitelist plugins under safe mode.
# Note that `some-other-jekyll-plugin` is not listed here. Therefore,

> #   it will not be loaded under safe mode.

Reviewer 2: Delete extra space in this comment - it does not appear to affect rendering.

whitelist:
  - jekyll-gist
  - jekyll-coffeescript
  - jekyll-seo-tag
```

In the absence of a Gemfile, one must manually ensure that listed plugins have been installed prior to invoking Jekyll. For example, the
latest versions of gems in the above list may be installed to a system-wide location by running:

```sh
gem install jekyll-gist jekyll-coffeescript jekyll-remote-theme some-other-jekyll-plugin
```
## Using a Gemfile

> The maintenance of various gem dependencies may be greatly simplified by using a Gemfile (usually at the root of site's source) in conjunction

Reviewer 2: "...usually at the root of the site's source"

> with a Rubygem named `bundler`. The Gemfile however **should** list all the primary dependencies of your site, including Jekyll itself, not

Reviewer 2: "...a Rubygem named Bundler"

Proper nouns FTW.

Author: I'm choosing to leave this in lowercase because it references the actual gem-name (`bundler`) instead of the library.
However, the consequent mention has been corrected as requested.

Reviewer 2: > I'm choosing to leave this in lowercase because it references the actual gem-name (`bundler`) instead of the library.

Got it. I was thinking about the distinction between Bundler (the project) v `bundler` (the executable which does a thing). See also Perl and `perl`.

Carry on.

> just gem-based plugins of the site because bundler narrows the scope of installed gems to just *runtime dependencies* resolved by evaluating

Reviewer 2: See prior comment.

the Gemfile. For example:

```ruby
source "https://rubygems.org"

# Use the latest version.
gem "jekyll"

# The theme of current site, locked to a certain version.
gem "minima", "2.4.1"

# Plugins of this site loaded during a build with proper
# site configuration.
gem "jekyll-gist"
gem "jekyll-coffeescript"
gem "jekyll-seo-tag", "~> 1.5"
gem "some-other-jekyll-plugin"

# A dependency of a custom-plugin inside `_plugins` directory.
gem "nokogiri", "~> 1.11"
```

> To install all of the listed gems, simply run `bundle install`.

Reviewer 2: Awkward.

"To install all of the gems listed"
